drivers/gpio/Kconfig.sim

```diff
@@ -0,0 +1,17 @@
+# Simulated GPIO configuration options
+
+# Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+# SPDX-License-Identifier: Apache-2.0
+
+config GPIO_SIM
+	bool "Simulated GPIO driver"
```

Comment:
I suppose we need to decide whether to use simulator or emulator for these features.

I feel that emulator is a slightly better term, since we are not actually trying to model the internal workings of the device, but just make it appear to work in the same way from the outside.

If we used simulators, we would be trying to model the internal workings of each device, e.g. a SPICE simulation of the workings of a PMIC.

So I think emulator makes more sense.

Reply:
From https://stackoverflow.com/a/1584701/236848:

> Emulation is the process of mimicking the outwardly observable behavior to match an existing target. The internal state of the emulation mechanism does not have to accurately reflect the internal state of the target which it is emulating.
>
> Simulation, on the other hand, involves modeling the underlying state of the target. The end result of a good simulation is that the simulation model will emulate the target which it is simulating.

I agree, emulation is the correct term here, and I will adjust my own usage.  A goal from my perspective is to be able to synthesize behavior corresponding to faults to ensure they're detected and handled properly.

Reply:
Except maybe not.... This driver *is* intended to model the internal workings of a GPIO peripheral, in that it maintains per-pin state and configuration and generates interrupts where a real device would.

For a temperature sensor code that simply records a specific temperature provided by a test framework and generates I2C responses that match what would be generated by a real sensor could be an emulator.  If the code instead implemented an algorithm to automatically adjust the temperature with the passage of time, generating threshold events as it changes, might be better described as a simulator.

I suppose it depends on the fidelity of "internal state", and more importantly on whether internal *behavior* is also exposed.

More discussion is needed, probably in the context of #27531.

Reply:
Yes it depends. You could argue that a real simulation of the GPIO would include voltage levels and model its reaction to the the drain and pull-ups, etc.

But since it doesn't simulate any particular GPIO chip or SoC GPIO block, and it doesn't really make sense to simulate a 'generic' GPIO, perhaps emulator is better?

Reply:
Going with "emulate", since that seems to already be the decision made in upstream.